service/registration.go

```diff
@@ -12,77 +17,283 @@ type RegistrationWrapper interface {
+		// credentials before the update routine is started.
+		credsWatcher:       make(chan Credentials, 1),
+		clientCredsChan:    make(chan Credentials),
+		ready:              make(chan struct{}),
+		mutex:              &sync.Mutex{},
+		currentCredentials: nil,
+		host:               host,
+		hostJSON:           hostJSON,
+		service:            service,
+		serviceJSON:        serviceJSON,
 	}
-	go r.worker(ctx)
-	return r
+
+	r.wg.Add(2)
+	go func() {
+		defer r.wg.Done()
+		r.currentCredentialsUpdater(ctx)
+	}()
+
+	go func() {
+		defer r.wg.Done()
+		initialID := r.initialRegistration(ctx)
+
+		if host.Public {
+			r.wg.Add(1)
+			go func() {
+				defer r.wg.Done()
+				r.credentialChangeWatcher(ctx, initialID)
+			}()
+		}
+		r.updateRoutine(ctx)
+	}()
+
+	return r, nil
+}
+
+func (r *Registration) Stop() error {
+	r.cancel()
+	r.wg.Wait()
+	return r.shutdownErr
 }
 
 // WaitRegistration wait for the first registration to happen, meaning that the service is succesfulled registred to the etcd service
-func (w *Registration) WaitRegistration() {
-	if w.Ready() {
-		return
-	}
-	<-w.readyChan
+func (r *Registration) WaitRegistration() {
+	<-r.ready
 }
 
 // Ready is a non blocking method that return true if the service is registred to the etcd service false otherwise
-func (w *Registration) Ready() bool {
-	w.mutex.Lock()
-	ready := w.ready
-	w.mutex.Unlock()
-	return ready
+func (r *Registration) Ready() bool {
+	select {
+	case <-r.ready:
+		return true
+	default:
+		return false
+	}
 }
 
 // UUID of the current host
-func (w *Registration) UUID() string {
-	return w.uuid
+func (r *Registration) UUID() string {
+	return r.host.UUID
 }
 
 // Credentials return the service credentials or an error if the service is not registred yet
-func (w *Registration) Credentials() (Credentials, error) {
-	w.mutex.Lock()
-	cred := w.curCredentials
-	w.mutex.Unlock()
+func (r *Registration) Credentials() (Credentials, error) {
+	r.mutex.Lock()
+	cred := r.currentCredentials
+	r.mutex.Unlock()
 	if cred == nil {
 		return Credentials{}, errors.New("Not ready")
 	}
 	return *cred, nil
 }
 
-func (w *Registration) worker(ctx context.Context) {
+func (r *Registration) currentCredentialsUpdater(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case newCredentials := <-r.clientCredsChan:
+			r.mutex.Lock()
+			r.currentCredentials = &newCredentials
+			r.mutex.Unlock()
+
+			if !r.Ready() {
+				close(r.ready)
+			}
+		}
+	}
+}
+
+func (r *Registration) initialRegistration(ctx context.Context) uint64 {
+	var (
+		id  uint64
+		err error
+	)
+	// First register on ETCD the service and the host, retry until done
+	// this operation can't fail.
+	//
+	// id is the current modification index of the service key.
+	// this is used for the watcher.
+	for {
+		id, err = r.registerService()
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Second)
+	}
+
+	for err != nil {
```

Comment:
after the previous loop, `err` equals `nil`. Will it ever enter this loop? Not sure.